pinot-query-planner/src/main/java/org/apache/pinot/query/planner/nodes/MailboxSendNode.java

```diff
@@ -0,0 +1,55 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.planner.nodes;
+
+import java.util.Collections;
+import java.util.List;
+import org.apache.calcite.rel.RelDistribution;
+
+
+public class MailboxSendNode extends AbstractStageNode {
```

Comment:
Based on the [discussion thread in design doc](https://docs.google.com/document/d/1iqzk54prgwIK59V3ELGoANs6h4v6ypHJkqMfiUhoI_I/edit?disco=AAAAWdz_Rfo), I think we should have an abstraction of ExchangeNode. ExchangeNode should encapsulate sender and receiver node. 

Similarly, there should be an abstraction for sender and receiver themselves.

Something like following....

Exchange
   BroadcastExchange
   SingleMergeExchange
   HashPartitionExchange

Sender
   BroadcastSender
   SingleSender
   HashPartitionSender

Receiver
   OrderedReceiver
   UnorderedReceiver

BroadcastExchange encapsulates
   BroadcastSender
   SomeReceiver

HashPartitionExchange encapsulates
   HashPartitionSender
   SomeReceiver


So ideally MailboxSend and MailboxReceive should be modeled as sender and receiver abstractions respectively as opposed to concrete implementations imo

Reply:
Yes I agree that we need to add more attributes to the stage nodes.
should we consider start simple and add attributes to the `ExchangeNode`? 

to me the only thing we need to separate is `SendExchangeNode` and `ReceiveExchangeNode`. the items you mentioned above can be inferred by the `Exchange.Type`
e.g. a `SendExchangeNode` with `Exchange.Type == BROADCAST` result in a broadcastSender and correspondingly the receiving side result in an UnorderedReceiver.